File: pkg/dart2cpp/sample/dart/basic_syntax.dart
/// 基础语法测试用例
/// 
/// 测试Dart基础语法特性，包括：
/// - 变量声明和数据类型
/// - 运算符
/// - 控制流语句
/// - 字符串操作
/// - 基本函数

void main() {
  print('🔥 基础语法测试开始');
  
  // 1. 基本数据类型测试
  testBasicTypes();
  
  // 2. 变量声明测试
  testVariableDeclarations();
  
  // 3. 运算符测试
  testOperators();
  
  // 4. 控制流测试
  testControlFlow();
  
  // 5. 字符串操作测试
  testStringOperations();
  
  // 6. 函数测试
  testFunctions();
  
  print('✅ 基础语法测试完成');
}

/// 测试基本数据类型
void testBasicTypes() {
  print('\n📌 测试基本数据类型');
  
  // 整数类型
  int intValue = 42;
  int hexValue = 0xFF;
  int negativeInt = -10;
  int zero = 0;
  
  // 浮点数类型
  double doubleValue = 3.14159;
  double scientificValue = 1.42e5;
  double negativeDouble = -2.5;
  
  // 布尔类型
  bool trueValue = true;
  bool falseValue = false;
  
  // 字符串类型
  String singleQuote = 'Hello';
  String doubleQuote = "World";
  String emptyString = '';
  String multiLine = '''
    This is a
    multi-line string
  ''';
  
  // 字符串插值
  String interpolation = 'The answer is ${intValue}';
  String expression = 'Sum: ${intValue + doubleValue}';
  
  print('  整数: $intValue, 十六进制: $hexValue, 负数: $negativeInt, 零: $zero');
  print('  浮点数: $doubleValue, 科学计数法: $scientificValue, 负数: $negativeDouble');
  print('  布尔值: $trueValue, $falseValue');
  print('  字符串: $singleQuote $doubleQuote');
  print('  插值: $interpolation');
  print('  表达式: $expression');
  print('  多行字符串长度: ${multiLine.length}');
}

/// 测试变量声明
void testVariableDeclarations() {
  print('\n📌 测试变量声明');
  
  // var 声明（类型推断）
  var autoInt = 100;
  var autoString = 'auto';
  var autoDouble = 3.14;
  var autoBool = true;
  
  // final 声明（运行时常量）
  final finalValue = 'cannot change';
  final int typedFinal = 200;
  final DateTime now = DateTime.now();
  
  // const 声明（编译时常量）
  const constValue = 'compile time constant';
  const double pi = 3.14159;
  const int maxValue = 100;
  
  // 可空类型
  int? nullableInt;
  String? nullableString = null;
  double? nullableDouble;
  
  // 非空断言和空值合并
  String nonNull = nullableString ?? 'default';
  nullableInt ??= 42;
  
  print('  var: $autoInt, $autoString, $autoDouble, $autoBool');
  print('  final: $finalValue, $typedFinal');
  print('  const: $constValue, $pi, $maxValue');
  print('  nullable: $nullableInt, $nullableString, $nullableDouble');
  print('  non-null: $nonNull');
}

/// 测试运算符
void testOperators() {
  print('\n📌 测试运算符');
  
  int a = 10, b = 3;
  double x = 5.5, y = 2.2;
  
  // 算术运算符
  print('  算术运算:');
  print('    ${a} + ${b} = ${a + b}');
  print('    ${a} - ${b} = ${a - b}');
  print('    ${a} * ${b} = ${a * b}');
  print('    ${a} / ${b} = ${a / b}');
  print('    ${a} % ${b} = ${a % b}');
  print('    ${a} ~/ ${b} = ${a ~/ b}');
  print('    -${a} = ${-a}');
  
  // 比较运算符
  print('  比较运算:');
  print('    ${a} == ${b}: ${a == b}');
  print('    ${a} != ${b}: ${a != b}');
  print('    ${a} > ${b}: ${a > b}');
  print('    ${a} < ${b}: ${a < b}');
  print('    ${a} >= ${b}: ${a >= b}');
  print('    ${a} <= ${b}: ${a <= b}');
  
  // 逻辑运算符
  bool p = true, q = false;
  print('  逻辑运算:');
  print('    ${p} && ${q}: ${p && q}');
  print('    ${p} || ${q}: ${p || q}');
  print('    !${p}: ${!p}');
  print('    !${q}: ${!q}');
  
  // 位运算符
  int m = 12, n = 5; // 1100, 0101
  print('  位运算:');
  print('    ${m} & ${n}: ${m & n}');
  print('    ${m} | ${n}: ${m | n}');
  print('    ${m} ^ ${n}: ${m ^ n}');
  print('    ~${m}: ${~m}');
  print('    ${m} << 1: ${m << 1}');
  print('    ${m} >> 1: ${m >> 1}');
  
  // 赋值运算符
  int c = 5;
  c += 2;
  print('  赋值运算: c += 2 结果: $c');
  c -= 1;
  print('  赋值运算: c -= 1 结果: $c');
  c *= 3;
  print('  赋值运算: c *= 3 结果: $c');
  c ~/= 2;
  print('  赋值运算: c ~/= 2 结果: $c');
  
  // 条件运算符
  String result = a > b ? 'a is greater' : 'b is greater';
  print('  三元运算符: $result');
  
  // 空值合并运算符
  String? nullable;
  String safe = nullable ?? 'default value';
  print('  空值合并: $safe');
}

/// 测试控制流
void testControlFlow() {
  print('\n📌 测试控制流');
  
  // if-else 语句
  int score = 85;
  print('  if-else 测试:');
  if (score >= 90) {
    print('    成绩: 优秀');
  } else if (score >= 80) {
    print('    成绩: 良好');
  } else if (score >= 70) {
    print('    成绩: 中等');
  } else {
    print('    成绩: 需要努力');
  }
  
  // switch-case 语句
  String grade = 'B';
  print('  switch-case 测试:');
  switch (grade) {
    case 'A':
      print('    等级: 优秀');
      break;
    case 'B':
      print('    等级: 良好');
      break;
    case 'C':
      print('    等级: 中等');
      break;
    default:
      print('    等级: 未知');
  }
  
  // for 循环
  print('  for 循环测试:');
  for (int i = 0; i < 5; i++) {
    print('    索引: $i, 值: ${i * i}');
  }
  
  // for-in 循环
  List<String> fruits = ['apple', 'banana', 'orange'];
  print('  for-in 循环测试:');
  for (String fruit in fruits) {
    print('    水果: $fruit');
  }
  
  // while 循环
  print('  while 循环测试:');
  int count = 0;
  while (count < 3) {
    print('    计数: $count');
    count++;
  }
  
  // do-while 循环
  print('  do-while 循环测试:');
  int num = 0;
  do {
    print('    数字: $num');
    num++;
  } while (num < 3);
  
  // break 和 continue
  print('  break 和 continue 测试:');
  for (int i = 0; i < 10; i++) {
    if (i == 2) continue;
    if (i == 7) break;
    print('    处理: $i');
  }
}

/// 测试字符串操作
void testStringOperations() {
  print('\n📌 测试字符串操作');
  
  String str1 = 'Hello';
  String str2 = 'World';
  String str3 = '  Dart Programming  ';
  
  // 字符串连接
  String concat = str1 + ' ' + str2;
  print('  连接: $concat');
  
  // 字符串插值
  int value = 42;
  String interpolated = 'The answer is $value';
  String expression = 'Sum: ${10 + 20}';
  print('  插值: $interpolated');
  print('  表达式插值: $expression');
  
  // 字符串属性
  print('  字符串属性:');
  print('    长度: ${str1.length}');
  print('    是否为空: ${"".isEmpty}');
  print('    是否不为空: ${str1.isNotEmpty}');
  
  // 字符串方法
  print('  字符串方法:');
  print('    大写: ${str1.toUpperCase()}');
  print('    小写: ${str2.toLowerCase()}');
  print('    去空格: "${str3.trim()}"');
  print('    子字符串: ${str1.substring(0, 4)}');
  print('    包含: ${str1.contains("ell")}');
  print('    开始于: ${str1.startsWith("He")}');
  print('    结束于: ${str1.endsWith("lo")}');
  print('    索引: ${str1.indexOf("l")}');
  print('    替换: ${str1.replaceAll("l", "L")}');
  print('    分割: ${concat.split(" ")}');
  
  // 字符串比较
  print('  字符串比较:');
  print('    相等: ${str1 == "Hello"}');
  print('    比较: ${str1.compareTo(str2)}');
  
  // 多行字符串
  String multiLine = '''
    第一行
    第二行
    第三行
  ''';
  print('  多行字符串行数: ${multiLine.split('\n').length}');
}

/// 测试函数
void testFunctions() {
  print('\n📌 测试函数');
  
  // 普通函数调用
  int sum = add(5, 3);
  print('  加法函数: add(5, 3) = $sum');
  
  // 可选位置参数
  greet('Alice');
  greet('Bob', 'Mr.');
  
  // 命名参数
  createUser(name: 'Charlie', age: 25);
  createUser(name: 'David', age: 30, email: 'david@example.com');
  
  // 匿名函数
  var multiply = (int a, int b) => a * b;
  print('  匿名函数: multiply(4, 5) = ${multiply(4, 5)}');
  
  // 箭头函数
  var square = (int x) => x * x;
  print('  箭头函数: square(6) = ${square(6)}');
  
  // 高阶函数
  var numbers = [1, 2, 3, 4, 5];
  var doubled = numbers.map((n) => n * 2).toList();
  print('  高阶函数 map: $doubled');
  
  var evens = numbers.where((n) => n % 2 == 0).toList();
  print('  高阶函数 where: $evens');
  
  // 函数作为参数
  int result = calculate(10, 5, add);
  print('  函数作为参数: calculate(10, 5, add) = $result');
  
  result = calculate(10, 5, subtract);
  print('  函数作为参数: calculate(10, 5, subtract) = $result');
}

// ============================================================================
// 辅助函数定义
// ============================================================================

/// 加法函数
int add(int a, int b) {
  return a + b;
}

/// 减法函数
int subtract(int a, int b) {
  return a - b;
}

/// 问候函数（可选位置参数）
void greet(String name, [String? title]) {
  if (title != null) {
    print('  问候: Hello, $title $name!');
  } else {
    print('  问候: Hello, $name!');
  }
}

/// 创建用户函数（命名参数）
void createUser({required String name, required int age, String? email}) {
  print('  创建用户: $name, $age岁${email != null ? ', 邮箱: $email' : ''}');
}

/// 计算函数（函数作为参数）
int calculate(int a, int b, int Function(int, int) operation) {
  return operation(a, b);
}

File: pkg/dart2cpp/sample/cpp_generated/basic_syntax.cpp
#include "dart2cpp.h"

// 工具宏定义

Nullable testBasicTypes();
Nullable testVariableDeclarations();
Nullable testOperators();
Nullable testControlFlow();
Nullable testStringOperations();
Nullable testFunctions();
Int add(Int a, Int b);
Int subtract(Int a, Int b);
Nullable greet(String name, String title = String(Null));
Nullable createUser(String name = String(Null), Int age = Int(Null), String email = String(Null));
Int calculate(Int a, Int b, std::function<Int(Int, Int)> operation);
Nullable testBasicTypes() {
  dart_print(dart_string("\n📌 测试基本数据类型"));
auto intValue = dart_int(42);
auto hexValue = dart_int(255);
auto negativeInt = dart_int(-10);
auto zero = dart_int(0);
auto doubleValue = dart_double(3.14159);
auto scientificValue = dart_double(142000.0);
auto negativeDouble = dart_double(-2.5);
auto trueValue = dart_bool(true);
auto falseValue = dart_bool(false);
auto singleQuote = dart_string("Hello");
auto doubleQuote = dart_string("World");
auto emptyString = dart_string("");
auto multiLine = dart_string("    This is a\n    multi-line string\n  ");
auto interpolation = dart_string("The answer is ") + (intValue).toString();
auto expression = dart_string("Sum: ") + ((intValue + doubleValue)).toString();
dart_print(dart_concat(dart_string("  整数: "), (intValue).toString(), dart_string(", 十六进制: "), (hexValue).toString(), dart_string(", 负数: "), (negativeInt).toString(), dart_string(", 零: "), (zero).toString()));
dart_print(dart_concat(dart_string("  浮点数: "), (doubleValue).toString(), dart_string(", 科学计数法: "), (scientificValue).toString(), dart_string(", 负数: "), (negativeDouble).toString()));
dart_print(dart_concat(dart_string("  布尔值: "), (trueValue).toString(), dart_string(", "), (falseValue).toString()));
dart_print(dart_concat(dart_string("  字符串: "), (singleQuote).toString(), dart_string(" "), (doubleQuote).toString()));
dart_print(dart_string("  插值: ") + (interpolation).toString());
dart_print(dart_string("  表达式: ") + (expression).toString());
dart_print(dart_string("  多行字符串长度: ") + (multiLine.size()).toString());
return Void;
}

Nullable testVariableDeclarations() {
  dart_print(dart_string("\n📌 测试变量声明"));
auto autoInt = dart_int(100);
auto autoString = dart_string("auto");
auto autoDouble = dart_double(3.14);
auto autoBool = dart_bool(true);
const auto finalValue = dart_string("cannot change");
const auto typedFinal = dart_int(200);
const auto now = ObjectPtr<DateTime>(new DateTime());
Int nullableInt(Null);
auto nullableString = Null;
Double nullableDouble(Null);
auto nonNull = dart_null_coalesce(nullableString, dart_string("default"));
if (dart_is_null(nullableInt)) nullableInt = dart_int(42);
dart_print(dart_concat(dart_string("  var: "), (autoInt).toString(), dart_string(", "), (autoString).toString(), dart_string(", "), (autoDouble).toString(), dart_string(", "), (autoBool).toString()));
dart_print(dart_concat(dart_string("  final: "), (finalValue).toString(), dart_string(", "), (typedFinal).toString()));
dart_print(dart_string("  const: compile time constant, 3.14159, 100"));
dart_print(dart_concat(dart_string("  nullable: "), (nullableInt).toString(), dart_string(", "), (nullableString).toString(), dart_string(", "), (nullableDouble).toString()));
dart_print(dart_string("  non-null: ") + (nonNull).toString());
return Void;
}

Nullable testOperators() {
  dart_print(dart_string("\n📌 测试运算符"));
auto a = dart_int(10);
auto b = dart_int(3);
auto x = dart_double(5.5);
auto y = dart_double(2.2);
dart_print(dart_string("  算术运算:"));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" + "), (b).toString(), dart_string(" = "), ((a + b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" - "), (b).toString(), dart_string(" = "), ((a - b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" * "), (b).toString(), dart_string(" = "), ((a * b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" / "), (b).toString(), dart_string(" = "), ((a / b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" % "), (b).toString(), dart_string(" = "), ((a % b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" ~/ "), (b).toString(), dart_string(" = "), (a.truncatingDivision(b)).toString()));
dart_print(dart_concat(dart_string("    -"), (a).toString(), dart_string(" = "), (a.operator_negate()).toString()));
dart_print(dart_string("  比较运算:"));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" == "), (b).toString(), dart_string(": "), ((a == b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" != "), (b).toString(), dart_string(": "), (!((a == b))).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" > "), (b).toString(), dart_string(": "), ((a > b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" < "), (b).toString(), dart_string(": "), ((a < b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" >= "), (b).toString(), dart_string(": "), ((a >= b)).toString()));
dart_print(dart_concat(dart_string("    "), (a).toString(), dart_string(" <= "), (b).toString(), dart_string(": "), ((a <= b)).toString()));
auto p = dart_bool(true);
auto q = dart_bool(false);
dart_print(dart_string("  逻辑运算:"));
dart_print(dart_concat(dart_string("    "), (p).toString(), dart_string(" && "), (q).toString(), dart_string(": "), (p && q).toString()));
dart_print(dart_concat(dart_string("    "), (p).toString(), dart_string(" || "), (q).toString(), dart_string(": "), (p || q).toString()));
dart_print(dart_concat(dart_string("    !"), (p).toString(), dart_string(": "), (!(p)).toString()));
dart_print(dart_concat(dart_string("    !"), (q).toString(), dart_string(": "), (!(q)).toString()));
auto m = dart_int(12);
auto n = dart_int(5);
dart_print(dart_string("  位运算:"));
dart_print(dart_concat(dart_string("    "), (m).toString(), dart_string(" & "), (n).toString(), dart_string(": "), (m.operator_bitwise_and(n)).toString()));
dart_print(dart_concat(dart_string("    "), (m).toString(), dart_string(" | "), (n).toString(), dart_string(": "), (m.operator_bitwise_or(n)).toString()));
dart_print(dart_concat(dart_string("    "), (m).toString(), dart_string(" ^ "), (n).toString(), dart_string(": "), (m.operator_bitwise_xor(n)).toString()));
dart_print(dart_concat(dart_string("    ~"), (m).toString(), dart_string(": "), (m.operator_bitwise_not()).toString()));
dart_print(dart_concat(dart_string("    "), (m).toString(), dart_string(" << 1: "), (m.operator_shift_left(dart_int(1))).toString()));
dart_print(dart_concat(dart_string("    "), (m).toString(), dart_string(" >> 1: "), (m.operator_shift_right(dart_int(1))).toString()));
auto c = dart_int(5);
c = (c + dart_int(2));
dart_print(dart_string("  赋值运算: c += 2 结果: ") + (c).toString());
c = (c - dart_int(1));
dart_print(dart_string("  赋值运算: c -= 1 结果: ") + (c).toString());
c = (c * dart_int(3));
dart_print(dart_string("  赋值运算: c *= 3 结果: ") + (c).toString());
c = c.truncatingDivision(dart_int(2));
dart_print(dart_string("  赋值运算: c ~/= 2 结果: ") + (c).toString());
auto result = (a > b) ? dart_string("a is greater") : dart_string("b is greater");
dart_print(dart_string("  三元运算符: ") + (result).toString());
String nullable(Null);
auto safe = dart_null_coalesce(nullable, dart_string("default value"));
dart_print(dart_string("  空值合并: ") + (safe).toString());
return Void;
}

Nullable testControlFlow() {
  dart_print(dart_string("\n📌 测试控制流"));
auto score = dart_int(85);
dart_print(dart_string("  if-else 测试:"));
if ((score >= dart_int(90))) {
dart_print(dart_string("    成绩: 优秀"));
} else {
if ((score >= dart_int(80))) {
dart_print(dart_string("    成绩: 良好"));
} else {
if ((score >= dart_int(70))) {
dart_print(dart_string("    成绩: 中等"));
} else {
dart_print(dart_string("    成绩: 需要努力"));
}
}
}
auto grade = dart_string("B");
dart_print(dart_string("  switch-case 测试:"));
if (grade == dart_string("A")) {
  dart_print(dart_string("    等级: 优秀"));
} else if (grade == dart_string("B")) {
  dart_print(dart_string("    等级: 良好"));
} else if (grade == dart_string("C")) {
  dart_print(dart_string("    等级: 中等"));
}else {
  dart_print(dart_string("    等级: 未知"));
}
dart_print(dart_string("  for 循环测试:"));
for (auto i = dart_int(0); (i < dart_int(5)); ++i) {
dart_print(dart_concat(dart_string("    索引: "), (i).toString(), dart_string(", 值: "), ((i * i)).toString()));
}
auto fruits = dart_literal(dart_string("apple"), dart_string("banana"), dart_string("orange"));
dart_print(dart_string("  for-in 循环测试:"));
auto sync_for_iterator = fruits->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto fruit = sync_for_iterator->next();
dart_print(dart_string("    水果: ") + (fruit).toString());
}
dart_print(dart_string("  while 循环测试:"));
auto count = dart_int(0);
while ((count < dart_int(3))) {
dart_print(dart_string("    计数: ") + (count).toString());
count = (count + dart_int(1));
}
dart_print(dart_string("  do-while 循环测试:"));
auto num = dart_int(0);
do {
dart_print(dart_string("    数字: ") + (num).toString());
num = (num + dart_int(1));
} while (((num < dart_int(3))).toBool());
dart_print(dart_string("  break 和 continue 测试:"));
for (auto i = dart_int(0); (i < dart_int(10)); ++i) {
if ((i == dart_int(2))) {
break;
}
if ((i == dart_int(7))) {
break;
}
dart_print(dart_string("    处理: ") + (i).toString());
};
return Void;
}

Nullable testStringOperations() {
  dart_print(dart_string("\n📌 测试字符串操作"));
auto str1 = dart_string("Hello");
auto str2 = dart_string("World");
auto str3 = dart_string("  Dart Programming  ");
auto concat = ((str1 + dart_string(" ")) + str2);
dart_print(dart_string("  连接: ") + (concat).toString());
auto value = dart_int(42);
auto interpolated = dart_string("The answer is ") + (value).toString();
auto expression = dart_string("Sum: ") + ((dart_int(10) + dart_int(20))).toString();
dart_print(dart_string("  插值: ") + (interpolated).toString());
dart_print(dart_string("  表达式插值: ") + (expression).toString());
dart_print(dart_string("  字符串属性:"));
dart_print(dart_string("    长度: ") + (str1.size()).toString());
dart_print(dart_string("    是否为空: ") + (dart_string("").isEmpty()).toString());
dart_print(dart_string("    是否不为空: ") + (str1.isNotEmpty()).toString());
dart_print(dart_string("  字符串方法:"));
dart_print(dart_string("    大写: ") + (str1.toUpperCase()).toString());
dart_print(dart_string("    小写: ") + (str2.toLowerCase()).toString());
dart_print(dart_concat(dart_string("    去空格: \""), (str3.trim()).toString(), dart_string("\"")));
dart_print(dart_string("    子字符串: ") + (str1.substring(dart_int(0), dart_int(4))).toString());
dart_print(dart_string("    包含: ") + (str1.contains(dart_string("ell"))).toString());
dart_print(dart_string("    开始于: ") + (str1.startsWith(dart_string("He"))).toString());
dart_print(dart_string("    结束于: ") + (str1.endsWith(dart_string("lo"))).toString());
dart_print(dart_string("    索引: ") + (str1.indexOf(dart_string("l"))).toString());
dart_print(dart_string("    替换: ") + (str1.replaceAll(dart_string("l"), dart_string("L"))).toString());
dart_print(dart_string("    分割: ") + (concat.split(dart_string(" "))).toString());
dart_print(dart_string("  字符串比较:"));
dart_print(dart_string("    相等: ") + ((str1 == dart_string("Hello"))).toString());
dart_print(dart_string("    比较: ") + (str1.compareTo(str2)).toString());
auto multiLine = dart_string("    第一行\n    第二行\n    第三行\n  ");
dart_print(dart_string("  多行字符串行数: ") + (multiLine.split(dart_string("\n"))->size()).toString());
return Void;
}

Nullable testFunctions() {
  dart_print(dart_string("\n📌 测试函数"));
auto sum = add(dart_int(5), dart_int(3));
dart_print(dart_string("  加法函数: add(5, 3) = ") + (sum).toString());
greet(dart_string("Alice"), String(Null));
greet(dart_string("Bob"), dart_string("Mr."));
createUser(dart_string("Charlie"), dart_int(25), String(Null));
createUser(dart_string("David"), dart_int(30), dart_string("david@example.com"));
auto multiply = makeFunction([&](Int a, Int b) { return (a * b); });
dart_print(dart_string("  匿名函数: multiply(4, 5) = ") + (multiply->apply(std::vector<Any>{dart_int(4), dart_int(5)})).toString());
auto square = makeFunction([&](Int x) { return (x * x); });
dart_print(dart_string("  箭头函数: square(6) = ") + (square->apply(std::vector<Any>{dart_int(6)})).toString());
auto numbers = dart_literal(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
auto doubled = numbers->map(makeFunction([&](Int n) { return (n * dart_int(2)); }))->toList();
dart_print(dart_string("  高阶函数 map: ") + (doubled).toString());
auto evens = numbers->where(makeFunction([&](Int n) { return ((n % dart_int(2)) == dart_int(0)); }))->toList();
dart_print(dart_string("  高阶函数 where: ") + (evens).toString());
auto result = calculate(dart_int(10), dart_int(5), makeFunction(&add));
dart_print(dart_string("  函数作为参数: calculate(10, 5, add) = ") + (result).toString());
result = calculate(dart_int(10), dart_int(5), makeFunction(&subtract));
dart_print(dart_string("  函数作为参数: calculate(10, 5, subtract) = ") + (result).toString());
}

Int add(Int a, Int b) {
  return (a + b);
}

Int subtract(Int a, Int b) {
  return (a - b);
}

Nullable greet(String name, String title) {
  if (!(dart_is_null(title))) {
dart_print(dart_concat(dart_string("  问候: Hello, "), (title).toString(), dart_string(" "), (name).toString(), dart_string("!")));
} else {
dart_print(dart_concat(dart_string("  问候: Hello, "), (name).toString(), dart_string("!")));
};
return Void;
}

Nullable createUser(String name, Int age, String email) {
  dart_print(dart_concat(dart_string("  创建用户: "), (name).toString(), dart_string(", "), (age).toString(), dart_string("岁"), (!(dart_is_null(email)) ? dart_string(", 邮箱: ") + (email).toString() : dart_string("")).toString()));
return Void;
}

Int calculate(Int a, Int b, std::function<Int(Int, Int)> operation) {
  return operation->apply(std::vector<Any>{a, b});
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("🔥 基础语法测试开始"));
testBasicTypes();
testVariableDeclarations();
testOperators();
testControlFlow();
testStringOperations();
testFunctions();
dart_print(dart_string("✅ 基础语法测试完成"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
